### PR TITLE
Us64T111 Implement editPost functionality

### PR DESCRIPTION
**Pull request by fmehdad97**
The way that the activity knows which post it is editing is by passing in the postID through .putExtra in the intent.

Maybe some of the editing functionality might be better to be refactored to the User model to make the activity leaner? I am open to suggestions and discussion from reviewer.

### CODE REVIEW

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java]: Unused here I think

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java:29]: Can you add an example here, showing how to call the activity with the putExtra? Do we need to send just postId, or also folderId?

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java:55]: I noticed that we don't have editTextPostTags hooked up to anything; should we? What's the desired functionality look like? Should we add an endpoint and include this in API calls? Deferring to you for your take on this

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java]: I like the chaining functionality here, but I wonder if a User method `getPost(String postId)` or `getPost(String folderId, String postId)` might be a simpler implementation? 

Since Posts are unique by ID, we don't *really* need to know the folder at this point; we could simply search for it through all folders. This is more performant, but the difference should be very small; up to you. But, abstracting the method call away might be nice. It would also give a way to handle a null folder OR post (because it can just return null if either post OR folder is null)

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java:65]: I like the null check here! You might consider also checking if getFolder returns null; otherwise, a NullPointerException will crash the app here

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java:81]: Ah, I was hoping for an example, thank you. This might be beneficial in the docstring too

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java:85]: is there a way to exit the activity?

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java]: Might be beneficial to revisit this and pull the error that the server is returning to the client; while 'not found' is one possible cause, there could be others (token expired, not authorized or no access, etc)

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/EditPostActivity.java]: I was reading on this yesterday and the onFailure here is called when the Response itself fails; could be due to connectivity or client-side failure, too, but generally means the connection to the server never happened

File: app/src/main/java/com/example/poste/activities/LoginActivity.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/activities/LoginActivity.java]: I think I might recommend uncommenting these lines and, if necessary, allowing the app to crash (preferably with logging) so we can determine why it crashes and how to resolve. Otherwise this sort of obfuscates the issue and control flow.

Not expecting you to debug / solve this in THIS pull request, but I'm weary of introducing unrelated control flow changes

File: app/src/main/java/com/example/poste/http/EditPostRequest.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/http/EditPostRequest.java:3]: Are we doing anything with tags?

File: app/src/main/java/com/example/poste/http/MyApiService.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/http/MyApiService.java:27]: Nicely done.

File: app/src/main/java/com/example/poste/models/Folder.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/models/Folder.java:66]: Nicely done. You might consider placing a method on User, perhaps two with one overloaded with FolderID; that's a matter of preference and abstraction level though.

File: app/src/main/java/com/example/poste/models/User.java
Comment (mccodav1) [inline app/src/main/java/com/example/poste/models/User.java:148]: Perhaps a copy-paste mismatch, this docstring doesn't appear to match the method

**Review by mccodav1 — APPROVED** [review]
I have a handful of suggestions / comments, but nothing I consider blocking on this PR. You might consider implementing some of these before merging over; alternatively, perhaps create User Stories in the backlog to resolve individual issues.

In industry it would not be uncommon to turn comments / suggestions here into user stories to be utilized in a future sprint, and might help us with sprint planning if we build them proactively

Well done

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java]: Has since been optimized, good catch! leftover from copying code and auto-import.

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java:29]: I have included the example.

Current implementation needs both folderId and postId.
I agree it is possible to do it with only postId... For some reason it just didn't occur to me when I was coding this

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java:55]: So the Tags was there from the prototype. I made the layout following the proto, and then during implementation I found tags wasnt implemented anywhere else... So I left it hanging for now. We can definitely incorporate it now, or after MVP when we begin to implement additional features such as searching for folders by name or posts by title, description or tags. It is based on a brief conversation in chat we had regarding adding searchable tags in the program. The tags could be user input, but my thought was it would probably be an ML classification algo. We can discuss more?

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java]: I can do the search based on only postID... I'll implement it ASAP.
But Im confused on what you mean by abstracting the method call away. Can you elaborate? I'll shoot a message over Slack.

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java:81]: Has since been added, thanks to your review

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java:85]: Cancel should take me back to the folder. I wasn't sure how to do that without the app crashing since the folder refactor hadnt been done yet. I'll clean that up and test it and commit again

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java:65]: currentFolder == null checks if getFolder returns null, since currentFolder gets the return value assigned to it from getFolder...
Unless I'm missing something?

File: app/src/main/java/com/example/poste/http/EditPostRequest.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/http/EditPostRequest.java:3]: Not at the moment. It's a future feature, I will comment//hide it in the layout for now

File: app/src/main/java/com/example/poste/models/Folder.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/models/Folder.java:66]: Partially done, can expand in future

File: app/src/main/java/com/example/poste/models/User.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/models/User.java:148]: Yes, fixed it since

File: app/src/main/java/com/example/poste/activities/LoginActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/LoginActivity.java]: Sounds good. I will uncomment and let it crash then

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java]: I've made the message more ambigious since to be more accurate

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java]: Added. 
I did not think of those errors before. So I did not include those errors, but it has since been abstracted to show the error message sent by the server

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java:65]: Sorry, I think I was unclear.

In line 54, we getStringExtra(); what if this value isn't passed? Is it possible to be null? Or rather, what happens if we pass an invalid value to `getFolder`?

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java:55]: I see. Just curious. No worries.

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java]: I simply meant calling something like `user.getPostById(<postid>)` and handling all the logic there, or calling later functions there if necessary

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (mccodav1) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java]: We should be able to parse out the error message from the server response directly and show that (or use it to determine what toast to make, at the least) -- I only see "Post to edit not found" or "Edit failed due to server error". 

Incidentally, the onFailure means a _request_ error, which likely indicates a connectivity error (means no reply was received from the server at all)

File: app/src/main/java/com/example/poste/activities/LoginActivity.java
Comment (mccodav1) [reply app/src/main/java/com/example/poste/activities/LoginActivity.java]: Best practice might be to create a user story to resolve, including specifics on this issue and what's causing it. That backlog of bugs will help us in this sprint.

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java]: Has since been implemented

File: app/src/main/java/com/example/poste/activities/EditPostActivity.java
Comment (fmehdad97) [reply app/src/main/java/com/example/poste/activities/EditPostActivity.java:65]: OHH yes that makes sense. I'll make a note of it to get to that ASAP.

**Pull request merged**
